### PR TITLE
[ColorPicker] First Orientation implementation

### PR DESCRIPTION
**Pull request by chingucoding**
## Description

First draft of orientation property implementation for ColorPicker. Design is not final and open for changes, awaiting feedback from the Designer here.

## Motivation and Context
First preliminary implementation of ColorPicker Orientation.

## How Has This Been Tested?

Tested manually.

## Screenshots (if appropriate):

V1 Style:
Screenshot of horizontal ColorPicker in V1 style

V2 Style:
Screenshot of horizontal ColorPicker in V2 style

### CODE REVIEW

File: dev/ColorPicker/ColorPicker.idl [hunk 1]
```diff
@@ -45,6 +45,8 @@ unsealed runtimeclass ColorPicker : Windows.UI.Xaml.Controls.Control
     [MUX_DEFAULT_VALUE("winrt::ColorSpectrumComponents::HueSaturation")]
     ColorSpectrumComponents ColorSpectrumComponents { get; set; };
     event Windows.Foundation.TypedEventHandler<ColorPicker, ColorChangedEventArgs> ColorChanged;
+    [MUX_DEFAULT_VALUE("winrt::Orientation::Vertical")]
```
Comment (ranjeshj) [inline dev/ColorPicker/ColorPicker.idl:48]: Can you add this under MUX_PUBLIC_V2  to keep it ABI compatible ? [https://github.com/microsoft/microsoft-ui-xaml/blob/0db5d0398cb38d69b06e26ef734cbbdebdebe774/idl/Microsoft.UI.Xaml.idl#L179](https://github.com/microsoft/microsoft-ui-xaml/blob/0db5d0398cb38d69b06e26ef734cbbdebdebe774/idl/Microsoft.UI.Xaml.idl#L179)

@jevansaks as FYI to double check.

File: dev/ColorPicker/ColorPicker.cpp [hunk 2]
```diff
@@ -428,6 +433,19 @@ void ColorPicker::OnColorSpectrumComponentsChanged(winrt::DependencyPropertyChan
...
+{
+    if (newOrientation == winrt::Orientation::Horizontal)
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Horizontal", true);
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Vertical", true);
+    }
+    UpdateLayout();
```
Comment (ranjeshj) [inline dev/ColorPicker/ColorPicker.cpp:446]: >UpdateLayout [](http://example.com/codeflow?start=4&length=12)

Is this required ?

File: dev/ColorPicker/ColorPicker.xaml [hunk 3]
```diff
@@ -10,122 +10,13 @@
...
-        <Setter Property="MinWidth" Value="312" />
         <Setter Property="IsTabStop" Value="False" />
         <contract7Present:Setter Property="CornerRadius" Value="{ThemeResource ControlCornerRadius}" />
         <Setter Property="Template">
             <Setter.Value>
                 <ControlTemplate TargetType="controls:ColorPicker">
-                    <Grid Background="{TemplateBinding Background}">
+                    <Grid x:Name="RootGrid" Background="{TemplateBinding Background}">
                         <Grid.Resources>
-                            <Style x:Key="ColorPickerSliderStyle" TargetType="primitives:ColorPickerSlider">
```
Comment (ranjeshj) [inline dev/ColorPicker/ColorPicker.xaml:22]: @StephenLPeters @karenbtlai do you know if there was a reason for having a custom slider style here ?

File: dev/ColorPicker/ColorPicker_v1.xaml [hunk 4]
```diff
@@ -327,161 +308,280 @@
...
-                                    <StackPanel x:Name="AlphaPanel" Orientation="Horizontal" Margin="0,12,0,0">
-                                        <TextBox x:Name="AlphaTextBox" Width="120" MaxLength="4" Text="100%"/>
-                                        <TextBlock x:Name="AlphaLabel" Text="Opacity" VerticalAlignment="Center" Margin="8,0,0,0"/>
-                                    </StackPanel>
-                                </StackPanel>
-                                <TextBox x:Name="HexTextBox" Grid.Column="1" MaxLength="7" Text="#FFFFFF" Margin="0,12,0,0" Width="132" HorizontalAlignment="Right" VerticalAlignment="Top" />
+
+                                <Grid.ColumnDefinitions>
+                                    <ColumnDefinition Width="{ThemeResource SliderPreContentMargin}" />
+                                    <ColumnDefinition Width="Auto" />
```
Comment (ranjeshj) [inline dev/ColorPicker/ColorPicker_v1.xaml:568]: Please ensure that the v1 styes work as expected. The test coverage for v1 styles is very limited now.

File: dev/ColorPicker/ColorPicker.idl (hunk 1 above)
Comment (jevansaks) [reply dev/ColorPicker/ColorPicker.idl:48]: Yes please. @ranjeshj do you have an ETA on the ABI compatibility test addition? Maybe now that WinUI 2.6 is out the door we can take a little time for that test?

We also need to change the framework package versioning scheme for 2.7 (there's a github issue tracking that, if there's a way for you to tag that on the "do this soon" that would be great. 🙂)

Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.idl]: Done!

File: dev/ColorPicker/ColorPicker.cpp (hunk 2 above)
Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.cpp]: Nope, removed it.

File: dev/ColorPicker/ColorPicker_v1.xaml (hunk 4 above)
Comment (chingucoding) [reply dev/ColorPicker/ColorPicker_v1.xaml:568]: Did manual testing with the v1 style and it seemed to work correctly.

File: dev/ColorPicker/ColorPicker.xaml [hunk 5]
```diff
@@ -254,9 +145,81 @@
...
+                                            <Setter.Value>
+                                                <RotateTransform Angle="180"/>
+                                            </Setter.Value>
+                                        </Setter>
+                                    
+                                        <Setter Target="MoreEntriesPanel.(Grid.Column)" Value="3"/>
+                                        <Setter Target="MoreButton.Margin" Value="0,0,0,12"/>
+                                    </VisualState.Setters>
+                                </VisualState>
+                                <VisualState x:Name="Vertical">
```
Comment (karenbtlai) [inline dev/ColorPicker/ColorPicker.xaml:181]: There should always be one orientation that is the default visual state, and not require any additional setters.

File: dev/ColorPicker/ColorPicker.cpp [hunk 6]
```diff
@@ -301,6 +302,10 @@ void ColorPicker::OnPropertyChanged(winrt::DependencyPropertyChangedEventArgs co
     {
         OnColorSpectrumComponentsChanged(args);
     }
+    else if (property == s_OrientationProperty)
+    {
+        UpdateOrientation(unbox_value<winrt::Orientation>(args.NewValue()));
```
Comment (karenbtlai) [inline dev/ColorPicker/ColorPicker.cpp:307]: I would unbox args.NewValue() in the UpdateOrientation function for consistency's sake.

File: dev/ColorPicker/ColorPicker.xaml (hunk 3 above)
Comment (karenbtlai) [reply dev/ColorPicker/ColorPicker.xaml:22]: It's for the slider to have the gradient colours rather than just a solid bar with the accent colour.

File: dev/ColorPicker/ColorPicker_v1.xaml [hunk 7]
```diff
@@ -238,8 +145,82 @@
...
+                                            <Setter.Value>
+                                                <RotateTransform Angle="180"/>
+                                            </Setter.Value>
+                                        </Setter>
+
+                                        <Setter Target="MoreEntriesPanel.(Grid.Column)" Value="3"/>
+                                        <Setter Target="MoreButton.Margin" Value="0,0,0,12"/>
+                                    </VisualState.Setters>
+                                </VisualState>
+                                <VisualState x:Name="Vertical">
```
Comment (karenbtlai) [inline dev/ColorPicker/ColorPicker_v1.xaml:181]: Same comment as the v2 styles.

File: dev/ColorPicker/ColorPicker.xaml (hunk 5 above)
Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.xaml:181]: Is that beneficial here? This would mean that we would need to introduce setters whose whole purpose is to undo the "default" settings (e.g. Grid.Row).

File: dev/ColorPicker/ColorPicker.cpp (hunk 6 above)
Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.cpp:307]: The thing is that we also want to be able to take an `Orientation` to update the orientation in `OnApplyTemplate` to be able to take the property set through XAML.

File: dev/ColorPicker/ColorPicker.xaml (hunk 5 above)
Comment (karenbtlai) [reply dev/ColorPicker/ColorPicker.xaml:181]: Yes. This will make it more readable as well.

File: dev/ColorPicker/ColorPicker.cpp (hunk 6 above)
Comment (karenbtlai) [reply dev/ColorPicker/ColorPicker.cpp:307]: UpdateOrientation will fire onLoad already, so you can actually remove it from OnApplyTemplate I'm pretty sure. You should also name the function to `OnOrientationChanged` for consistency's sake as well.

File: dev/ColorPicker/ColorPicker.xaml (hunk 5 above)
Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.xaml]: Updated the code now :)

File: dev/ColorPicker/ColorPicker.cpp (hunk 6 above)
Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.cpp]: Updated that now, if we don't need it than it's best to remove that code. Thank you for pointing that out.

**Comment by gabbybilka** [comment]
Hi @chingucoding,

I've just cloned your branch to get a screenshot of the horizontal ColorPicker in the Collapsable states (showing More/Less button) but wasn't able to change the control width to view all of the elements at once. See below:
image

image

Changing the control width in the MUX test app to get a better screenshot crashed the app or didn't work as intended. Do you have any ideas on what the problem could be? Thanks!

**Comment by chingucoding** [comment]
Hi @gabbybilka if the app crashes when you changed something, that is a big issue. If you know how to reproduce this, that would be great so we can fix this.

Regarding taking a screenshot, you can uncheck the "Render innerframe in lab dimensions" button as shown in the screenshot below:

Screenshot of app with "Render innerframe in lab dimensions" button marked

File: dev/ColorPicker/ColorPicker.cpp [hunk 8]
```diff
@@ -301,6 +301,10 @@ void ColorPicker::OnPropertyChanged(winrt::DependencyPropertyChangedEventArgs co
     {
         OnColorSpectrumComponentsChanged(args);
     }
+    else if (property == s_OrientationProperty)
+    {
+        OnUpdateOrientation(args);
```
Comment (karenbtlai) [inline dev/ColorPicker/ColorPicker.cpp:306]: The naming convention is On*Changed, so I would change this to `OnOrientationChanged`

**Comment by karenbtlai** [comment]
>     UpdateVisualState(true /* useTransitions */);

I would move this bit into UpdateVisualState and simply call UpdateVisualState. This keeps all the VisualState transitions consistently in one place.

---
Refers to: dev/ColorPicker/ColorPicker.cpp:366 in d85b627. [](commit_id = d85b6273183a22452b03212270a71e1a427ffceb, deletion_comment = False)

File: dev/ColorPicker/ColorPicker.xaml [hunk 9]
```diff
@@ -254,9 +147,72 @@
...
                             </VisualStateGroup>
+                            <VisualStateGroup x:Name="Orientation">
+                                <VisualState x:Name="Vertical"/>
+                                <VisualState x:Name="Horizontal">
+                                    <VisualState.Setters>
+                                        <Setter Target="RootGrid.MinHeight" Value="{StaticResource ColorPickerVerticalOrientationMinHeight}"/>
+                                        <Setter Target="RootGrid.MaxHeight" Value="{StaticResource ColorPickerVerticalOrientationMaxHeight}"/>
+                                        <Setter Target="RootGrid.MinWidth" Value="0"/>
+                                        <Setter Target="RootGrid.MaxWidth" Value="10000"/>
+                                        <Setter Target="ColorSpectrumGrid.Margin" Value="0,0,0,0"/>
```
Comment (karenbtlai) [inline dev/ColorPicker/ColorPicker.xaml:158]: nit: change to `0`, as well as the other `0,0,0,0` below.

**Comment by gabbybilka** [comment]
@chingucoding 
> Hi @gabbybilka if the app crashes when you changed something, that is a big issue. If you know how to reproduce this, that would be great so we can fix this.
> 
To repro the crash, change the orientation to horizontal then change the control width from -1 to any other value. If I change the control width first, then change the orientation to horizontal, it is ok.


> Regarding taking a screenshot, you can uncheck the "Render innerframe in lab dimensions" button as shown in the screenshot below:
>

Selecting this worked, thanks! A note, if you change the control width with the slider in horizontal mode, the max width is 392 which cuts off content.

File: dev/ColorPicker/ColorPicker.xaml [hunk 10]
```diff
@@ -341,163 +297,298 @@
...
+                    <Grid>
+                        <Grid.Resources>
+                            <Style x:Key="SliderThumbStyle" TargetType="Thumb">
+                                <Setter Property="BorderThickness" Value="1"/>
+                                <Setter Property="Background" Value="{ThemeResource ColorPickerSliderThumbBackground}"/>
+                                <Setter Property="Template">
+                                    <Setter.Value>
+                                        <ControlTemplate TargetType="Thumb">
+                                            <Border BorderBrush="{ThemeResource SliderThumbBorderBrush}"
+                                                                            BorderThickness="{TemplateBinding BorderThickness}"
```
Comment (karenbtlai) [inline dev/ColorPicker/ColorPicker.xaml:479]: nit: fix indentation

**Comment by chingucoding** [comment]
> > ```
> > UpdateVisualState(true /* useTransitions */);
> > ```
> 
> I would move this bit into UpdateVisualState and simply call UpdateVisualState. This keeps all the VisualState transitions consistently in one place.
> 
> Refers to: dev/ColorPicker/ColorPicker.cpp:366 in [d85b627](https://github.com/microsoft/microsoft-ui-xaml/commit/d85b6273183a22452b03212270a71e1a427ffceb). [](commit_id = [d85b627](https://github.com/microsoft/microsoft-ui-xaml/commit/d85b6273183a22452b03212270a71e1a427ffceb), deletion_comment = False)

Not sure which code exactly  I should move into UpdateVisualState. The logic to determine what orientation state to go to?

**Comment by chingucoding** [comment]
@gabbybilka I think the issues should be fixed now :)

**Comment by karenbtlai** [comment]
> > > ```
> > > UpdateVisualState(true /* useTransitions */);
> > > ```
> > 
> > 
> > I would move this bit into UpdateVisualState and simply call UpdateVisualState. This keeps all the VisualState transitions consistently in one place.
> > Refers to: dev/ColorPicker/ColorPicker.cpp:366 in [d85b627](https://github.com/microsoft/microsoft-ui-xaml/commit/d85b6273183a22452b03212270a71e1a427ffceb). [](commit_id = [d85b627](https://github.com/microsoft/microsoft-ui-xaml/commit/d85b6273183a22452b03212270a71e1a427ffceb), deletion_comment = False)
> 
> Not sure which code exactly I should move into UpdateVisualState. The logic to determine what orientation state to go to?

Oops, it seems CodeFlow referred to the wrong part of the code. Let me redo that!

File: dev/ColorPicker/ColorPicker.cpp [hunk 11]
```diff
@@ -428,6 +432,18 @@ void ColorPicker::OnColorSpectrumComponentsChanged(winrt::DependencyPropertyChan
...
+void ColorPicker::OnOrientationChanged(winrt::DependencyPropertyChangedEventArgs const& args)
+{
+    if (unbox_value<winrt::Orientation>(args.NewValue()) == winrt::Orientation::Horizontal)
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Horizontal", true);
+    }
+    else
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Vertical", true);
+    }
```
Comment (karenbtlai) [inline dev/ColorPicker/ColorPicker.cpp:444]: I would move this bit into UpdateVisualState and simply call UpdateVisualState. This keeps all the VisualState transitions consistently in one place.

Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.cpp]: Sure thing, updated the code now. Good idea!

**Comment by ranjeshj** [comment]
@chingucoding @karenbtlai the hang in the colorpicker tests seem to be because in CanSelectColorFromSpectrumRTL it looks like the colorpicker is clipped and we are tapping in an area of the spectrum that is clipped. Also, the waiters seem to wait indefinitely. If there is a way to timeout we should do that so the entire test suite doesn't hang.

image

**Comment by ranjeshj** [comment]
setting more button visible is currently hiding all the input boxes which is not expected.

File: dev/ColorPicker/ColorPicker.cpp [hunk 12]
```diff
@@ -37,7 +37,7 @@ void ColorPicker::OnApplyTemplate()
 
     m_alphaSlider.set(GetTemplateChildT<winrt::ColorPickerSlider>(L"AlphaSlider", thisAsControlProtected));
     m_alphaSliderGradientBrush.set(GetTemplateChildT<winrt::LinearGradientBrush>(L"AlphaSliderGradientBrush", thisAsControlProtected));
-    m_alphaSliderBackgroundRectangle.set(GetTemplateChildT<winrt::Rectangle>(L"AlphaSliderBackgroundRectangle", thisAsControlProtected));
+    m_alphaSliderBackgroundRectangle.set(GetTemplateChildT<winrt::Rectangle>(L"AlphaSliderGradientBackgroundRectangle", thisAsControlProtected));
```
Comment (ranjeshj) [inline dev/ColorPicker/ColorPicker.cpp:40]: did we rename a template part ? If someone has a previous template, this wont work anymore ?

Comment (chingucoding) [reply dev/ColorPicker/ColorPicker.cpp]: Yes, but those changes were not necessary so I reverted them since we shouldn't break customers without a good reason.

File: MUXControls.sln [hunk 13]
```diff
@@ -820,6 +820,7 @@ Global
 		dev\ScrollPresenter\ScrollPresenter.vcxitems*{ad0c90b0-4845-4d4b-88f1-86f653f8171b}*SharedItemsImports = 4
 		dev\ScrollView\ScrollView.vcxitems*{ad0c90b0-4845-4d4b-88f1-86f653f8171b}*SharedItemsImports = 4
 		dev\SplitButton\SplitButton.vcxitems*{ad0c90b0-4845-4d4b-88f1-86f653f8171b}*SharedItemsImports = 4
+		dev\SplitView\SplitView.vcxitems*{ad0c90b0-4845-4d4b-88f1-86f653f8171b}*SharedItemsImports = 4
```
Comment (karenbtlai) [inline MUXControls.sln:823]: Is this change intentional?

**Review by karenbtlai — APPROVED** [review]

File: MUXControls.sln (hunk 13 above)
Comment (chingucoding) [reply MUXControls.sln:823]: That was not intentional. no idea how that happened. @karenbtlai Would you like me to revert those changes?
